src/h5web/dataset-visualizer/DimensionMapper.tsx

```diff
@@ -1,86 +1,78 @@
   onChange: (d: DimensionMapping) => void;
 }
 
 function DimensionMapper(props: Props): JSX.Element {
-  const { activeVis, onChange } = props;
-  const { rawDims, mapping } = useVisProps();
+  const { activeVis, rawDims, mapperState, onChange } = props;
 
   if (activeVis === Vis.Line && rawDims.length === 2) {
     return (
       <div className={styles.mapper}>
         <div className={styles.buttonGroupWrapper}>
-          <span className={styles.buttonGroupLabel}>X</span>
+          <span className={styles.dimensionName}>x</span>
           <ButtonGroup
             ariaLabel="Dimension as X"
             className={styles.buttonGroup}
             buttonClassName={styles.btn}
             buttons={Object.keys(rawDims).map(dimKey => {
               const dimIndex = Number(dimKey);
               return {
-                label: dimKey,
-                isSelected: mapping.x === dimIndex,
+                label: `D${dimIndex}`,
+                isSelected: mapperState[dimIndex] === 'x',
                 onClick: () => {
-                  if (mapping.x !== dimIndex) {
-                    const newMapping: DimensionMapping = {
-                      x: dimIndex,
-                      slicingIndices: { [mapping.x]: 0 },
-                    };
-                    onChange(newMapping);
+                  const prevX = mapperState.indexOf('x');
+                  if (prevX !== dimIndex) {
+                    const newMapperState = mapperState.slice();
+                    newMapperState[prevX] = 0;
+                    newMapperState[dimIndex] = 'x';
+                    onChange(newMapperState);
                   }
                 },
               };
             })}
           />
         </div>
         <div className={styles.sliderWrapper}>
-          {Object.entries(mapping.slicingIndices).map(
-            ([dimNumber, slicingIndex]) => (
-              <>
-                <button
-                  type="button"
-                  className={styles.btn}
-                  aria-pressed="true"
-                >
-                  {dimNumber}
-                </button>
+          {mapperState.map((slicingIndex, dim) =>
+            typeof slicingIndex === 'number' ? (
+              // eslint-disable-next-line react/no-array-index-key
+              <Fragment key={dim}>
+                <span className={styles.sliderLabel}>D{dim}</span>
                 <ReactSlider
-                  key={dimNumber}
                   className={styles.slider}
                   trackClassName={styles.sliderTrack}
                   thumbClassName={styles.sliderThumb}
                   renderThumb={(thumbProps, state) => (
                     <div {...thumbProps}>{state.valueNow}</div>
                   )}
                   defaultValue={slicingIndex}
-                  onChange={values => {
-                    if (values === undefined) {
+                  onChange={value => {
+                    if (!isNumber(value)) {
```

Comment:
The stronger guard clause allows to enforce the fact that `value` is defined AND is a number.